Generate better docs when using pybind11

knarfS opened the pull request:
This two commits will generate nicer looking docs when documenting modules generated by pybind11 with no source code available:

* Use the `__pdoc__` Dict for documenting class variables.
* Parse the signature string in the first line of the docstring when available.

knarfS commented:
Heck, I have no idea why these tests are failing....

knarfS commented:
I have some minor updates to this PR. Can I squash and force push them or do you want them as separate commits?

Any ideas about the failing tests? Any hints what's wrong?

kernc commented:
Commits will be squashed and/or rebased at the end, so I don't particularly mind either way.

I'm not really certain about failures. One subtest seems to be failing:
https://travis-ci.org/pdoc3/pdoc/jobs/638395055#L290
https://github.com/pdoc3/pdoc/blob/afcadcd3a4b47b4859a531f33b385d87856fb950/pdoc/test/__init__.py#L1204-L1207

The other one
https://travis-ci.org/pdoc3/pdoc/jobs/638395056#L242
is related to typing, but I can't even find the supposed affected line. :open_mouth:

knarfS commented:
I've done some investigation on the failing tests:

- Job #403.1 and #403.2:
I don't even know what `/csv.ext` is or what those test are doing...

- Job #403.3:
I guess that is referring to line 1232 (not 1237):
https://github.com/pdoc3/pdoc/blob/6885ab530fe03e975730b556a6bf869dd5717026/pdoc/__init__.py#L1231-L1232
Haven't changed that in my commits....

@kernc: Any help on those failing tests is appreciated!

knarfS commented:
Finally all tests have passed!
Also rebased to current master.

kernc commented on pdoc/__init__.py:
```diff
@@ -1098,24 +1127,59 @@ def params(self, *, annotate: bool = False, link: Callable[[Doc], str] = None) -
-            # TODO: Extract signature from the first line of the docstring, i.e.
-            # https://github.com/mitmproxy/pdoc/commit/010d996003bc5b72fcf5fa515edbcc0142819919
-            return ["..."]
+            if doc_obj.obj.__doc__ is None:
```
Let's make this a separate function/staticmethod `_signature_from_docstring()`. Use cleaned `doc_obj.docstring` instead of `__doc__`?

knarfS replied:
I will test and change this.

kernc commented on pdoc/__init__.py:
```diff
@@ -1098,24 +1127,59 @@ def params(self, *, annotate: bool = False, link: Callable[[Doc], str] = None) -
+            if doc_obj.obj.__doc__ is None:
+                return ["..."]
+
+            # Extract signature from the first line of the docstring for C builtin function
```
Can you show some examples how the signature in docstring looks "in the wild"? It makes regexes easier to follow. Where in CPython is that stringified signature constructed?

kernc replied:
So, if the string signatures are typically valid Python code, I think it might be more manageable and far less intricate to reach signature object with something dumb like:
```py
my_locals = {}
exec('def ' + string_signature + ': pass', {}, my_locals)
inspect.signature(my_locals[doc_obj.name])
```

knarfS replied:
> Can you show some examples how the signature in docstring looks "in the wild"? It makes regexes easier to follow.

The signatures have to be valid function definitions (without the `def` keyword) and can be as simple as
```
foo(bar)
```
or fully fletched with type hints, etc.:
```
foo(bar1: str, bar2: int = 1) -> Tuple[int, str]
```
Here is a real world example from my `smuview` module. The signature was generated by pybind11:
```
add_user_channel(self: smuview.BaseDevice, channel_name: str, channel_group_name: str) -> smuview.UserChannel
```
> Where in CPython is that stringified signature constructed?

I doubt (but I don't know) if there is a CPython function, which generates the function signature. In my case this is done by [pybind11](https://pybind11.readthedocs.io/en/stable/advanced/misc.html#generating-documentation-using-sphinx) itself:
> By default, pybind11 automatically generates and prepends a signature to the docstring of a function [...]
  
The stringified signature is documented in PEP 257, [Capter "One-line Docstrings"](https://www.python.org/dev/peps/pep-0257/#one-line-docstrings):
> The one-line docstring should NOT be a "signature" reiterating the function/method parameters (which can be obtained by introspection).
> [...]
> This type of docstring is only appropriate for C functions (such as built-ins), where introspection is not possible.
  
And from the [numpy docstring guide](https://numpydoc.readthedocs.io/en/latest/format.html#sections):
> The function signature is normally found by introspection and displayed by the help function. For some functions (notably those written in C) the signature is not available, so we have to specify it as the first line of the docstring: [...]

knarfS replied:
> So, if the string signatures are typically valid Python code, I think it might be more manageable and far less intricate to reach signature object with something dumb like:
> 
> ```python
> my_locals = {}
> exec('def ' + string_signature + ': pass', {}, my_locals)
> inspect.signature(my_locals[doc_obj.name])
> ```

Great! A first test worked!
Much simpler, but you have to use the actual `locales()` to get the imported stuff of the module that should be documented. I don't know if this conflicts with something else, but I will test.

kernc replied:
Did you then manage to push this further? :smiley:

knarfS replied:
Yes, I've done more testing with `exec()` and `inspect.signature()` and run into a problem.
When you have python built-in types as type hints, this is working great. But when the signature contains types from other packages, it starts to get complicated.

kernc commented on pdoc/__init__.py:
```diff
@@ -1098,24 +1127,59 @@ def params(self, *, annotate: bool = False, link: Callable[[Doc], str] = None) -
+    def _params(doc_obj, annotate=False, link=None, module=None):
         try:
-            signature = inspect.signature(inspect.unwrap(func_obj))
+            signature = inspect.signature(inspect.unwrap(doc_obj.obj))
```
Since Python 3.5, `inspect.unwrap()` seems to be [not needed](https://docs.python.org/3.5/library/inspect.html#inspect.signature).

knarfS replied:
I will change this.

kernc commented on pdoc/__init__.py:
```diff
@@ -826,10 +828,25 @@ def __init__(self, name, module, obj, *, docstring=None):
+                if refname in pdoc.keys() and not pdoc_string:
+                    pdoc_string = pdoc[refname]
+                    if pdoc_string is False:
+                        continue
```
What are these two blocks? :eyes:

knarfS replied:
This is for retrieving the docstring of class variables from the `__pdoc__` Dict.

First lookup (block) is for the short reference name (e.g.: `class_name.variable_name`) the second is for the full reference name (`module_name.class_name.variable_name`), as one or the other could be used. When set to `False`, the variable will be excluded from the documentation, as defined [here](https://pdoc3.github.io/pdoc/doc/pdoc/#overriding-docstrings-with-__pdoc__).

kernc replied:
I don't get it. Are you saying that without this addition the described documented behavior doesn't, in fact, work? :open_mouth:

knarfS replied:
Yes you are correct: Entries for class variables in the `__pdoc__ ` Dict are not evaluated at the moment.

kernc replied:
Yikes. :sweat_smile: Would you open a separate PR for that?

knarfS replied:
Sure! Will do later today...

Should I remove commit f77f882 and rebase this PR? Will this conversation be lost when I rebase the PR?

kernc replied:
I think it gets obsoleted/collapsed once the touching lines are changed.

kernc replied:
Continued in https://github.com/pdoc3/pdoc/pull/162.

kernc commented on pdoc/__init__.py:
```diff
@@ -1196,6 +1212,36 @@ def __repr__(self):
+                if strings:
+                    my_locals, my_globals = {}, self.module.obj.__dict__
+                    try:
+                        exec('def {}: pass'.format(filter(strings[0])), my_globals, my_locals)
```
> https://github.com/pdoc3/pdoc/pull/148#discussion_r407103658

Updated proposal includes globals from wherever the object is defined. Would this work?

knarfS replied:
I ran into the same issue I mentioned above.
For example the signature
```
get_sample(self: smuview.AnalogSampleSignal, pos: int) -> Tuple[int, float]
```
will cause the error
```
NameError: name 'smuview' is not defined
```

Because `AnalogSampleSignal` is in `self.module.obj.__dict__`, but not `smuview.AnalogSampleSignal`, I've changed the signature to
```
get_sample(self: AnalogSampleSignal, pos: int) -> Tuple[int, float]
```
But then I get the error
```
NameError: name 'Tuple' is not defined
```

Adding `from typing import Tuple` to the exec string works until the next unknown type occurs.

kernc replied:
> I've changed the signature to

How did you do that? Does pybind11 generally refer to its exported members with a module prefix?

We can prepopulate passed globals with `typing` members. We can't catch them all cases, but sure can try a few. :D

knarfS replied:
Also ast (no type hints) and typed_ast (only type comments) are useless.
Only solutions I see atm is a regex parser (from my earlier commit) with the risk of not catching complex types. Or a new sophisticated (classic) parser for signatures.

knarfS replied:
> > I've changed the signature to
> 
> How did you do that?

Hacked into `_signature_from_string()` with `.replace('smuview.', '')` for quick'n'dirty testing ;)

> Does pybind11 generally refer to its exported members with a module prefix?

Yes, afaict.

knarfS replied:
Here is the documentation for the whole smuview module, generated with `pydoc.help(smuview)`. The docstrings are pretty much unformated: 
https://github.com/knarfS/smuview/blob/master/doc/smuview_python_bindings.txt

Here is the documentation generated by pdoc with my patches applied:
https://github.com/knarfS/smuview/blob/master/doc/smuview_python_bindings.html

HTH

kernc replied:
I think the latest pushed change should work with modules like yours.

What did you mean with ast, typed_ast?

knarfS replied:
> What did you mean with ast, typed_ast?

The [Python Abstract Syntax Tree](https://docs.python.org/3/library/ast.html) offers the possibility to parse Python code. Here is an example for function parameters: https://stackoverflow.com/a/51271704

Since version 3.8 ast supports [type comments](https://www.python.org/dev/peps/pep-0484/#type-comments), before 3.8 there is [typed_ast](https://github.com/python/typed_ast).
But none of them is supporting type hints/annotations, so they are useless for the function signatures (definitions see [here](https://github.com/pdoc3/pdoc/pull/148#discussion_r377598769)).

kernc commented on pdoc/__init__.py:
```diff
@@ -1196,6 +1212,36 @@ def __repr__(self):
+            signature = None
+            for expr, cleanup_docstring, filter in (
+                    # Full proper typed signature, such as one from pybind11
+                    (r'^{}\(.*\)(?: -> .*)?$', True, lambda s: s),
```
Will this RE work with pybind11? The [example in their documentation](https://pybind11.readthedocs.io/en/master/basics.html#default-arguments) indicates the signature docstring line looks like this?
```
Signature : (i: int = 1, j: int = 2) -> int

A function which adds two numbers ...
```

knarfS replied:
No, the signature docstring for that specific function in the example would be:
```
add(i: int = 1, j: int = 2) -> int
```

The example is a formated output (probably from pydoc.help()).

kernc replied:
I don't think `help` reformats. Must be an old doc example then. Thanks.

knarfS commented:
I've added the pybind11 special datatype `capsule` to the globals dict.

knarfS commented:
I've added a new branch (https://github.com/knarfS/pdoc/tree/pybind11_3), that is rebased to the current master and removed the obsolete "ENH: Use __pdoc__ Dict for class variables if available." commit. But not squashed together yet.

If you want to I can move that to this PR.

All tests so far have worked, but I'm not using any 3rd party libs/types in the function signatures (yet).

kernc commented:
Managed to solve the merging spectacle. :sweat_smile: Let's see how far this works with real projects. Thanks for all the work on this one! :cake:

knarfS commented:
Thanks, everything is working fine now!
